Stream server for exec, attach and port forward

jjlakis opened the pull request:
Here's my proposition of streaming server implementation.
I had to fulfill a lot of dependencies for newly-vendored `kubelet/server/streaming` package.
Right now, `ocic` can return endpoint URL, server is working and handles these requests (with error described below)
Issues/todo:
* Catch TTY requests
* Implement exec client in `ocic`

runcom commented:
@mrunalp @cyphar

mrunalp commented:
Will take a detailed look later today/tomorrow 👍 
Won't we need changes to conmon as well?
(I don't mind doing this for the short term to get to feature complete. )

jjlakis commented:
@mrunalp We will need probably. Not much details to see now :)

runcom commented on oci/oci.go:
```diff
@@ -42,6 +43,20 @@ func New(runtimePath string, containerDir string, conmonPath string, conmonEnv [
+	// TODO: Is it should be started somewhere else?
+	go func() {
+		r.streamServer.Start(true)
+	}()
```
We should make sure this goroutine has been really started before moving along. Like:
```go
ch := make(chan struct{}) 
go func() {
    close(ch)
    r.streamServer.Start(true)
}() 
<-ch
...
```

runcom replied:
BTW, to reply to your TODO, I believe it's fine starting the server here :)

runcom commented:
@jjlakis could you provide some details on how to test this?

jjlakis commented:
@runcom `ExecResponse` provides URL for HTTP streaming request (this is what I am reutrning with `ocic`). And for now, you can see how it is working by simple `curl`:
```
$ sudo ocic ctr exec --id $NEWPOD_ID ls
URL:
http://0.0.0.0:10101/exec/57ef0c541c824772314f7082b3da8154fb7cf600f689d90df82b7ee72a52d802?command=ls&error=1&output=1
$ curl "http://0.0.0.0:10101/exec/57ef0c541c824772314f7082b3da8154fb7cf600f689d90df82b7ee72a52d802?command=ls&error=1&output=1"
unable to upgrade: missing upgrade headers in request: http.Header{"User-Agent":[]string{"curl/7.47.0"}, "Accept":[]string{"*/*"}}
```
Looks like `kubelet` adds some additional info to http request.

runcom commented:
@jjlakis yeah I noticed the same by using curl 2 days ago, however, I was more interested in the real exec, meaning I wanted to see logs actually and curl doesn't seem to help there

cyphar commented:
@runcom Right now, it doesn't look like it can work with getting logs from `conmon` (given #162 is not merged).

I've committed the next 3 weeks towards working on this, so I'm very happy to work on the required conmon changes so that we can get attach, exec and other such things working inside `conmon`. Unfortunately that will block this PR because the streaming server necessarily can't work until `conmon` has the required streaming features.

jjlakis commented:
@mrunalp @cyphar. I refactored my code a bit and added [endpoint](https://github.com/kubernetes-incubator/cri-o/pull/282/files#diff-6c6158181ead3695a96cd6c6e257c683R29) for the http server to catch requests. Endpoint is directly calling "exec" so it doesn't need any changes in `conmon`. http server wants from the client to create streams, so `curl` cannot make it work, only kubelet can for now. 

After merging #189 to this PR, **I am able** to see logs, right now only for non-TTY request (such as `ls`), here's the branch: https://github.com/intelsdi-x/cri-o/tree/exec_rebase

I started looking on TTY requests now. I also think about two ways of using future `ocic exec`:
* `ocic exec <container> <command> --url` to see just an endpoint url (`ExecResponse` contents), working now
* `ocic exec <container> <command>` to actually exec, kubelet-like http client needs to be implemented here

cyphar commented:
@jjlakis 

> Endpoint is directly calling "exec" so it doesn't need any changes in conmon. http server wants from the client to create streams, so curl cannot make it work, only kubelet can for now.

While that is cool (and it is super awesome to see it working), it's a layer violation and IMO the right way of handling it is all through conmon. In particular, the issues you mention about TTYs is precisely because of how TTY allocation works with `runc`.

I will continue working on the conmon code for logging, and then I will spend some time trying to figure out a design for conmon that will work for streaming. My main concern is with the choice of language at the moment.

jjlakis commented:
I agree your concerns about layer violation, but I am wondering what more can `conmon` do than just calling `exec`. I am even not sure what is bad in directly calling `runc exec` (`execsync` do this). Moreover, we have to implement Exec() endpoint in the way I did (if we want to use kubelet packages of course instead of implementing from the beginning) to easily return `io.Writer` and `io.Reader`. Of course we can change binary that is called in endpoint to `conmon` or something like this, I just want to highlight that this soultion seems to be natural IMO and we should base on this instead of looking for sophisticated way to sending logs.

cyphar commented:
@jjlakis 

> @cyphar I agree your concerns about layer violation, but I am wondering what more can conmon do than just calling exec.

Conmon can hook exec sessions into the logging system.

> I am even not sure what is bad in directly calling runc exec (execsync do this). Moreover, we have to implement Exec() endpoint in the way I did (if we want to use kubelet packages of course instead of implementing from the beginning) to easily return io.Writer and io.Reader.

The main reason this is an issue is because it means that exec sessions will die as soon as the connection dies. This is /probably/ fine for exec sessions, since I don't think the CRI has a concept of attaching to exec sessions. However, it is an issue for the core sessions of a container in a pod.

So we have to implement a logging system that isn't like Docker's (read: it actually is sane), and because we have to implement it _anyway_ why not avoid the layer violation and bifocation of code that handles runC calls and implement it all in conmon?

We also get the possible extensibility benefits of:

1. Logging exec sessions.
2. Allowing to re-attach to exec sessions.
3. Not killing exec sessions if the connection is killed.

> Of course we can change binary that is called in endpoint to conmon or something like this, I just want to highlight that this soultion seems to be natural IMO and we should base on this instead of looking for sophisticated way to sending logs.

We have to implement the sophisticated code _anyway_, which is sort of my point. I have a feeling the reason your TTY code isn't working is because of how TTYs are created with runC -- which is something that `conmon` has code to handle (in #162) and which you would have to replicate in Go.

jjlakis commented:
@cyphar Thanks for all the input. Please keep me in loop when thinking about streaming design in conmon. I'll think about separate PR only with streaming client for `ocic`. Last question for now:

> The main reason this is an issue is because it means that exec sessions will die as soon as the connection dies. This is /probably/ fine for exec sessions, since I don't think the CRI has a concept of attaching to exec sessions. However, it is an issue for the core sessions of a container in a pod.

Could you provide an example of this core session?

cyphar commented:
@jjlakis 

> Thanks for all the input. Please keep me in loop when thinking about streaming design in conmon. I'll think about separate PR only with streaming client for ocic.

Will do. I will write up an issue for it / proposal as soon as #162 is done.

> Could you provide an example of this core session?

Sorry that was the wrong terminology. I was referring to the `kubectl attach` endpoints, which require you to be able to interact with the stdio of a container in a pod (while still logging it). This will have to be implemented in conmon, so from my PoV it makes sense to also implement `exec` support in `conmon`.

jjlakis commented:
@runcom See this comment: https://github.com/kubernetes-incubator/cri-o/pull/282#issuecomment-269604462 .

jjlakis commented:
FYI rebased. Thinking about kubelet-like streaming http client for ocic.
cc @runcom @cyphar

mrunalp commented:
@jjlakis I spent some time looking into this and I came to the same conclusion about implementing exec the easiest way as you have done in this PR. Can you please rebase? :) (We are planning to talk about this in the call today if you can join.)

jjlakis commented:
@mrunalp Sorry I missed the message yesterday, I can rebase it for sure :)

mrunalp commented:
@jjlakis No worries. Thanks! :)

mrunalp commented:
@jjlakis Is this rebased?

jjlakis commented:
@mrunalp Yes it is, but I am unable to test it on cluster.

jjlakis commented:
Pulled @sameo's branch

mrunalp commented:
@jjlakis Thanks! You still need to update the vendored dependencies. The build is failing without that.

sameo commented:
@mrunalp Should a simple `dep ensure -update` do the trick ?

mrunalp commented:
@sameo Some adjustments will be required. For e.g. go-restful/swagger is only present in older commit that is also vendored into k8s.

sameo commented:
@mrunalp yes, I hit that swagger issue when trying to revendor...

mrunalp commented:
@sameo This commit worked for me yesterday 09691a3b6378b740595c1002f40c34dd5f218a22

jjlakis commented:
Closing. https://github.com/kubernetes-incubator/cri-o/pull/466
